index.js

```diff
@@ -87,122 +32,213 @@ function busboy (options) {
+    })
 
-    const req = this.raw
+    request.pipe(bb)
 
-    const busboyOptions = deepmerge.all([{ headers: req.headers }, options || {}, opts || {}])
-    const stream = busboy(busboyOptions)
-    var completed = false
-    var files = 0
-    var count = 0
-    var callDoneOnNextEos = false
+    function onField (name, fieldValue, fieldnameTruncated, valueTruncated) {
+      // don't overwrite prototypes
+      if (getDescriptor(Object.prototype, name)) return
 
-    req.on('error', function (err) {
-      stream.destroy()
-      if (!completed) {
-        completed = true
-        done(err)
+      const value = {
+        fieldname: name,
+        value: fieldValue,
+        fieldnameTruncated,
+        valueTruncated,
+        fields: body
       }
-    })
 
-    stream.on('finish', function () {
-      log.debug('finished receiving stream, total %d files', files)
-      if (!completed && count === files) {
-        completed = true
-        setImmediate(done)
+      if (body[name] === undefined) {
+        body[name] = value
+      } else if (Array.isArray(body[name])) {
+        body[name].push(value)
       } else {
-        callDoneOnNextEos = true
+        body[name] = [body[name], value]
       }
-    })
-
-    stream.on('file', wrap)
 
-    req.pipe(stream)
-      .on('error', function (error) {
-        req.emit('error', error)
-      })
+      ch(value)
+    }
 
-    function wrap (field, file, filename, encoding, mimetype) {
-      log.debug({ field, filename, encoding, mimetype }, 'parsing part')
-      files++
-      eos(file, waitForFiles)
-      if (field === '__proto__') {
-        file.destroy(new Error('__proto__ is not allowed as field name'))
-        return
+    function onFile (name, file, filename, encoding, mimetype) {
+      const value = {
+        fieldname: name,
+        filename,
+        encoding,
+        mimetype,
+        file,
+        fields: body
       }
-      handler(field, file, filename, encoding, mimetype)
+      if (body[name] === undefined) {
+        body[name] = value
+      } else if (Array.isArray(body[name])) {
+        body[name].push(value)
+      } else {
+        body[name] = [body[name], value]
+      }
+
+      ch(value)
     }
 
-    function waitForFiles (err) {
-      if (err) {
-        completed = true
-        done(err)
-        return
-      }
+    function onError (err) {
+      lastError = err
+    }
+
+    function onEnd () {
+      cleanup()
+      bb.removeListener('finish', onEnd)
+      bb.removeListener('error', onEnd)
+      ch(lastError)
+    }
 
-      if (completed) {
-        return
+    function cleanup () {
+      // keep finish listener to wait all data flushed
+      // keep error listener to wait stream error
+      request.removeListener('close', cleanup)
+      bb.removeListener('field', onField)
+      bb.removeListener('file', onFile)
+      bb.removeListener('close', cleanup)
+    }
+
+    return parts
+  }
+
+  async function saveRequestFiles () {
```

Comment:
I think this should throw if some of the other methods are called first.

Reply:
Good point.